File: LCR/Solution_LCR_176.java

class Solution_LCR_176 {
    public static void main(String[] args) {
        Solution_LCR_176 solution = new Solution_LCR_176();

        // Test case 1: Balanced tree
        TreeNode root1 = new TreeNode(1,
                            new TreeNode(2,
                                new TreeNode(4),
                                new TreeNode(5)),
                            new TreeNode(3));
        System.out.println(solution.isBalanced(root1)); // Should print true

        // Test case 2: Unbalanced tree
        TreeNode root2 = new TreeNode(1,
                            new TreeNode(2,
                                new TreeNode(4,
                                    new TreeNode(6),
                                    null),
                                new TreeNode(5)),
                            new TreeNode(3));
        System.out.println(solution.isBalanced(root2)); // Should print false
    }

    public boolean isBalanced(TreeNode root) {
        return check(root) != -1;
    }

    private int check(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int L = check(root.left);
        if (L == -1) return -1;

        int R = check(root.right);
        if (R == -1) return -1;

        if (Math.abs(L - R) > 1) return -1;

        return Math.max(L, R) + 1;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(int val) {
        this.val = val;
    }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: LCR/Solution_LCR_176.py
class TreeNode :
    def __init__(self,val = 0 , left = None , right = None):
        self.val = val;
        self.left = left;
        self.right = right;

class Solution :
    def isBalanced(self,root:TreeNode) -> bool :
        def check(node):
            if not node:
                return 0;
            L = check(node.left);
            if L == -1 :
                return -1;
            R = check(node.right);
            if R == -1 :
                return -1;

            if abs(L - R) > 1 :
                return -1;

            return max(L,R) + 1;
        return check(root) != -1;


def main() :
    s = Solution();
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)
    root.left.left.left = TreeNode(8)
    root.left.left.right = TreeNode(9)
    print(s.isBalanced(root));


if __name__ == "__main__" :
    main()
